packages/liveblocks-react-ui/src/primitives/Composer/index.tsx

```diff
@@ -347,6 +320,175 @@ function ComposerEditorMentionSuggestionsWrapper({
   );
 }
 
+function ComposerEditorFloatingToolbarWrapper({
+  id,
+  position = FLOATING_TOOLBAR_POSITION,
+  dir,
+  FloatingToolbar,
+  hasFloatingToolbarRange,
+  setHasFloatingToolbarRange,
+}: ComposerEditorFloatingToolbarWrapperProps) {
+  const editor = useSlateStatic();
+  const { isFocused } = useComposer();
+  const { portalContainer } = useLiveblocksUIConfig();
+  const [contentRef, contentZIndex] = useContentZIndex();
+  const [isPointerDown, setPointerDown] = useState(false);
+  const isOpen = isFocused && !isPointerDown && hasFloatingToolbarRange;
+  const {
+    refs: { setReference, setFloating },
+    strategy,
+    isPositioned,
+    placement,
+    x,
+    y,
+  } = useFloatingWithOptions({
+    type: "range",
+    position,
+    dir,
+    alignment: "center",
+    open: isOpen,
+  });
+
+  useLayoutEffect(() => {
+    if (!isFocused) {
+      return;
+    }
+
+    const handlePointerDown = () => setPointerDown(true);
+    const handlePointerUp = () => setPointerDown(false);
+
+    document.addEventListener("pointerdown", handlePointerDown);
+    document.addEventListener("pointerup", handlePointerUp);
+
+    return () => {
+      document.removeEventListener("pointerdown", handlePointerDown);
+      document.removeEventListener("pointerup", handlePointerUp);
+    };
+  }, [isFocused]);
+
+  useOnComposerEditorChange(() => {
+    // Detach from previous selection range (if any) to avoid sudden jumps
+    setReference(null);
+
+    // Then, wait for the next render to ensure the selection is updated
+    requestAnimationFrame(() => {
+      const domSelection = window.getSelection();
+
+      // Finally, show the toolbar if there's a selection range
+      if (
+        !editor.selection ||
+        SlateRange.isCollapsed(editor.selection) ||
+        !domSelection ||
+        !domSelection.rangeCount
+      ) {
+        setHasFloatingToolbarRange(false);
+        setReference(null);
+      } else {
+        setHasFloatingToolbarRange(true);
+
+        const domRange = domSelection.getRangeAt(0);
+        setReference(domRange);
+      }
+    });
```

Comment:
Just to make sure I understand this correctly, we listen to changes to the editor (selection, state update, etc) and set the reference element of the floating toolbar to null and immediately in the next animation frame, based on whether a selection range is collapsed or not, we say if the floating toolbar should appear or not.  
And, the reason why we wait for the next animation frame is because the selection change is known only a frame after editor change occurs?

Reply:
We use the [`inline`](https://floating-ui.com/docs/inline) Floating UI middleware for positioning which uses [`getClientRects`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getClientRects) to check individual lines instead of the bounds but if we don't wait a frame before updating the reference, somehow the DOM selection returns an empty list from `getClientRects` when you toggle marks (and change the editor's markup). Because of that empty list, `inline` falls back to using `getBoundingClientRect` which makes the toolbar jump around on toggle.

https://github.com/user-attachments/assets/214ee522-58f1-4d8c-8596-0f4e156188ce